segregation/inference/comparative.py

```diff
@@ -0,0 +1,453 @@
+    else:
+        df2 = df2[[group_pop_var2, total_pop_var2]]
+
+    df1["group_composition"] = (df1[group_pop_var1] / df1[total_pop_var1]).fillna(0)
+    df2["group_composition"] = (df2[group_pop_var2] / df2[total_pop_var2]).fillna(0)
+
+    df1["compl_pop_var"] = df1[total_pop_var1] - df1[group_pop_var1]
+    df2["compl_pop_var"] = df2[total_pop_var2] - df2[group_pop_var2]
+
+    df1["compl_composition"] = (df1["compl_pop_var"] / df1[total_pop_var1]).fillna(0)
+    df2["compl_composition"] = (df2["compl_pop_var"] / df2[total_pop_var2]).fillna(0)
+
+    df1["counterfactual_group_pop"] = (
+        df1["group_composition"].rank(pct=True).apply(df2["group_composition"].quantile)
+        * df1[total_pop_var1]
+    )
+    df2["counterfactual_group_pop"] = (
+        df2["group_composition"].rank(pct=True).apply(df1["group_composition"].quantile)
+        * df2[total_pop_var2]
+    )
+
+    df1["counterfactual_compl_pop"] = (
+        df1["compl_composition"].rank(pct=True).apply(df2["compl_composition"].quantile)
+        * df1[total_pop_var1]
+    )
+    df2["counterfactual_compl_pop"] = (
+        df2["compl_composition"].rank(pct=True).apply(df1["compl_composition"].quantile)
+        * df2[total_pop_var2]
+    )
+
+    df1["counterfactual_total_pop"] = (
+        df1["counterfactual_group_pop"] + df1["counterfactual_compl_pop"]
+    )
+    df2["counterfactual_total_pop"] = (
+        df2["counterfactual_group_pop"] + df2["counterfactual_compl_pop"]
+    )
+
+    return df1, df2
+
+
+def sim_share(
+    df1, df2, group_pop_var1, total_pop_var1, group_pop_var2, total_pop_var2,
+):
+    """Simulate the spatial population distribution of a region using the CDF of a comparison region.
+
+    For each spatial unit i in region 1, take the unit's rank in the distribution, and swap the group share
+    with the value of the corresponding rank in region 2. The share is the minority population of unit i
+    divided by total population of minority population. This approach will shift the total population of
+    each unit without changing the regional proportion of each group
+
+    Parameters
+    ----------
+    df1 : pandas.DataFrame or geopandas.GeoDataFrame
+        dataframe for first dataset with columns holding group and total population counts
+    df2 : pandas.DataFrame or geopandas.GeoDataFrame
+        dataframe for second dataset with columns holding group and total population counts
+    group_pop_var1 : str
+        column holding population counts for group of interest on input df1
+    total_pop_var1 : str
+        column holding total population counts on input df1
+    group_pop_var2 : str
+        column holding population counts for group of interest on input df2
+    total_pop_var2 : str
+        column holding total population counts on input df2
+
+    Returns
+    -------
+    two pandas.DataFrame
+        dataframes with simulated population columns appended
+
+    """
+    df1 = df1.copy()
```

Comment:
I think that it might look better to have this pattern into a `_prepare()` function (for example, [`pointpats.distance_statistics`](https://github.com/pysal/pointpats/blob/master/pointpats/distance_statistics.py#L31))? That way, if you add steps to the preparation/validation step before these computations, you can do that once...

Reply:
good idea. I managed to pull a lot of this logic into segregation's [base classes](https://github.com/pysal/segregation/blob/403bcf6db748c366d9083ffc98cf66dfa1f536a4/segregation/_base.py#L87) during the 2.0 push, but was a bit less thorough on this pass